### PR TITLE
Adds Session endpoints model

### PR DESCRIPTION
**Pull request by shubham1g5**
[Spec](https://docs.google.com/document/d/111mch19MnKMcXf91h5dmjvomYnket5x696Io3yLEjcE/edit?ts=5f724532#heading=h.5h6283rsmrrl)

### CODE REVIEW

File: src/main/java/org/commcare/suite/model/Suite.java [hunk 1]
```diff
@@ -46,18 +46,20 @@
...
     }
 
     public Suite(int version, Hashtable<String, Detail> details,
-                 Hashtable<String, Entry> entries, Vector<Menu> menus) {
+                 Hashtable<String, Entry> entries, Vector<Menu> menus, Hashtable<String, Endpoint> endpoints) {
         this.version = version;
         this.details = details;
         this.entries = entries;
         this.menus = menus;
+        this.endpoints = endpoints;
```
Comment (ShivamPokhriyal) [inline src/main/java/org/commcare/suite/model/Suite.java:62]: Should this be `externalized`?

File: src/main/java/org/commcare/xml/EndpointParser.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    private static final String NAME_ARGUMENT = "argument";
+    private static final String NAME_STACK = "stack";
+
+    public EndpointParser(KXmlParser parser) {
+        super(parser);
+    }
+
+    @Override
+    public Endpoint parse() throws InvalidStructureException, IOException, XmlPullParserException, UnfullfilledRequirementsException {
+        String id = parser.getAttributeValue(null, "id");
```
Comment (ShivamPokhriyal) [inline src/main/java/org/commcare/xml/EndpointParser.java:26]: Can this `id` be null or empty?

**Review by ShivamPokhriyal — APPROVED** [review]
Looks good to me.

File: src/main/java/org/commcare/xml/EndpointParser.java (hunk 2 above)
Comment (shubham1g5) [reply src/main/java/org/commcare/xml/EndpointParser.java:26]: it should not be. I should add a validation here to that point. Thanks.

File: src/main/java/org/commcare/suite/model/Suite.java (hunk 1 above)
Comment (shubham1g5) [reply src/main/java/org/commcare/suite/model/Suite.java:62]: yes, totally missed out on it. Thanks for the catch

**Review by ShivamPokhriyal — DISMISSED** [review]

File: src/main/java/org/commcare/suite/model/StackFrameStep.java [hunk 3]
```diff
@@ -234,4 +236,13 @@ public void setType(String elementType) {
     public String getElementType() {
         return elementType;
     }
+
+    // find and replace any arguments $argument with their value in {@code arguments}
+    public void substituteArgument(HashMap<String, String> arguments) {
+        if (value.indexOf("$") != -1) {
```
Comment (ctsims) [inline src/main/java/org/commcare/suite/model/StackFrameStep.java:242]: I was imagining that we would provide arguments through the normal XPath engine. You can set variables in the Evaluation Context, and they will be evaluated as normal.

Munging the string in place can come with pretty complex issues, especially if the user inputs are complex

**Review by ctsims — CHANGES_REQUESTED** [review]
Made a note on the way we introduce and process arguemnts.

File: src/main/java/org/commcare/suite/model/StackFrameStep.java [hunk 4]
```diff
@@ -18,7 +18,9 @@
 import java.io.DataOutputStream;
 import java.io.IOException;
 import java.util.Enumeration;
+import java.util.HashMap;
```
Comment (ctsims) [inline src/main/java/org/commcare/suite/model/StackFrameStep.java:21]: Nit: These are no-ops now

File: src/main/java/org/commcare/xml/EndpointParser.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+import java.util.Vector;
+
+public class EndpointParser extends ElementParser<Endpoint> {
+
+    static final String NAME_ENDPOINT = "endpoint";
+    private static final String NAME_ARGUMENT = "argument";
+    private static final String NAME_STACK = "stack";
```
Comment (ctsims) [inline src/main/java/org/commcare/xml/EndpointParser.java:18]: Suggestion: Moving this ID to the StackOpsParser and re-using the static String between this and the other stack ops parsing contexts.

**Review by ctsims — APPROVED** [review]
Looks good! Optional suggestions in feedback.

**Review by ctsims — APPROVED** [review]
Looks great! Thanks again for all of the throughness on getting this implemented!

**Pull request merged**